text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,648 @@
+          IsImportedBucket: !result.statementAdded,
+        },
+      });
+    }
+
+    if (this.bucket.encryptionKey) {
+      const provider = S3OriginAccessControlKeyPolicyProvider.getOrCreateProvider(scope, S3_ORIGIN_ACCESS_CONTROL_KEY_RESOURCE_TYPE,
+        {
+          description: 'Lambda function that updates SSE-KMS key policy to allow CloudFront distribution access.',
+        });
+      provider.addToRolePolicy({
+        Action: ['kms:PutKeyPolicy', 'kms:GetKeyPolicy', 'kms:DescribeKey'],
+        Effect: 'Allow',
+        Resource: [this.bucket.encryptionKey.keyArn],
+      });
+
+      new CustomResource(scope, 'S3OriginKMSKeyPolicyCustomResource', {
+        resourceType: S3_ORIGIN_ACCESS_CONTROL_KEY_RESOURCE_TYPE,
+        serviceToken: provider.serviceToken,
+        properties: {
+          DistributionId: distributionId,
+          KmsKeyId: this.bucket.encryptionKey.keyId,
+          AccountId: this.bucket.env.account,
+          Partition: Stack.of(scope).partition,
+        },
+      });
+    }
+
+    const originBindConfig = super.bind(scope, options);
+
+    // Update configuration to set OriginControlAccessId property
+    return {
+      ...originBindConfig,
+      originProperty: {
+        ...originBindConfig.originProperty!,
+        originAccessControlId: this.originAccessControl.originAccessControlId,
+      },
+    };
+  }
+
+  /**
+   * If you're using origin access control (OAC) instead of origin access identity, specify an empty `OriginAccessIdentity` element.
+   * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-distribution-s3originconfig.html#cfn-cloudfront-distribution-s3originconfig-originaccessidentity
+   */
+  protected renderS3OriginConfig(): cloudfront.CfnDistribution.S3OriginConfigProperty | undefined {
+    return { originAccessIdentity: '' };
+  }
+}
+```
+
+#### `S3Origin` Construct Modifications
+
+The `S3Origin` constructor will need additional logic to determine how to configure the S3 origin (either as website endpoint, using OAI, or using OAC).
+
+```ts
+export class S3Origin implements cloudfront.IOrigin {
+  private readonly origin: cloudfront.IOrigin;
+
+  constructor(bucket: s3.IBucket, props: S3OriginProps = {}) {
+    if (props.originAccessControl && props.originAccessIdentity) {
+      throw new Error('Only one of originAccessControl or originAccessIdentity can be specified for an origin.');
+    }
+
+    if (bucket.isWebsite) {
+      this.origin = new HttpOrigin(bucket.bucketWebsiteDomainName, {
+        protocolPolicy: cloudfront.OriginProtocolPolicy.HTTP_ONLY, // S3 only supports HTTP for website buckets
+        ...props,
+      });
+    } else if (props.originAccessIdentity || !FeatureFlags.of(bucket.stack).isEnabled(cxapi.CLOUDFRONT_USE_ORIGIN_ACCESS_CONTROL)) {
+      this.origin = new S3BucketOrigin(bucket, props);
+    } else {
+      this.origin = new S3BucketOacOrigin(bucket, props);
```

Comment:
Using my above suggestions of `S3BucketOrigin` as an abstract class with two static methods this could look like:
```ts
if (bucket.isWebsite) {
      this.origin = new HttpOrigin(bucket.bucketWebsiteDomainName, {
        protocolPolicy: cloudfront.OriginProtocolPolicy.HTTP_ONLY, // S3 only supports HTTP for website buckets
        ...props,
      });
    } else if (props.originAccessIdentity || !FeatureFlags.of(bucket.stack).isEnabled(cxapi.CLOUDFRONT_USE_ORIGIN_ACCESS_CONTROL)) {
      this.origin = S3BucketOrigin.withAccessIdentity(bucket, props);
    } else {
      this.origin = S3BucketOrigin.withAccessControl(bucket, props);
```